src/deploy/functions/validate.ts

```diff
@@ -87,31 +91,27 @@ export function packageJsonIsValid(
 }
 
 export function checkForInvalidChangeOfTrigger(
-  fn: CloudFunctionTrigger,
-  exFn: CloudFunctionTrigger
+  fn: backend.FunctionSpec,
+  exFn: backend.FunctionSpec
 ) {
-  if (fn.httpsTrigger && !exFn.httpsTrigger) {
+  const wantEventTrigger = backend.isEventTrigger(fn.trigger);
+  const haveEventTrigger = backend.isEventTrigger(exFn.trigger);
+  if (!wantEventTrigger && haveEventTrigger) {
     throw new FirebaseError(
       `[${getFunctionLabel(
-        fn.name
+        fn
       )}] Changing from a background triggered function to an HTTPS function is not allowed. Please delete your function and create a new one instead.`
     );
   }
-  if (!fn.httpsTrigger && exFn.httpsTrigger) {
+  if (wantEventTrigger && !haveEventTrigger) {
     throw new FirebaseError(
       `[${getFunctionLabel(
-        fn.name
+        fn
       )}] Changing from an HTTPS function to an background triggered function is not allowed. Please delete your function and create a new one instead.`
     );
   }
-  if (fn.eventTrigger?.service != exFn.eventTrigger?.service) {
```

Comment:
Thanks for the explanation! Right when i joined, I spent some time debugging some GCF + CF3 tests that wouldn't work on staging, and I distinctly remember being super confused by why preprod talked to prod services.

Anecdotes aside, makes sense to me to deprecate this.

Reply:
In case you're curious, there's a valid argument for both a staging tier (across all services) vs a staging service (talking to prod).

The staging tier folks argue that it's the only way to make sure that dev/staging bugs don't wipe out prod data or overload prod with buggy traffic.

The staging service folks argue that a staging tier tells you nothing about whether you can actually push to prod (there may be dependencies on other features that are still in staging) and it's functionally no different from a graceful rollout.

I'm personally in #teamStagingService, but I recognize that Google has _mostly_ sided with #teamStagingTier.